scripts/rv_expression_association/get_rv_outliers.py

```diff
@@ -0,0 +1,197 @@
+    gene_name: str,
+    sample_mapping_file: str,  # 'scrna-seq/grch38_association_files/OneK1K_CPG_IDs.tsv'
+    gene_file: str,  # e.g., 'scrna-seq/grch38_association_files/gene_location_files/GRCh38_geneloc_chr1.tsv'
+    window_size: int,
+):
+    """for a given individual,
+    and gene for which that individual is an expression outlier
+    get all variants within a given window,
+
+    then filter for variants that:
+    - are alt (0/1 or 1/1) for that individual
+    - are biallelic SNVs
+    - have regulatory consequences based on VEP annotations
+
+    also annotate variants with the following:
+    - CADD score
+    - MAF within the OneK1K sample
+    - MAF in gnomad
+
+    create the following table:
+    donor ID | gene ID | variant ID | position | CADD | MAF (OneK1K) | MAF (gnomad) | ..
+    """
+
+    # file matching OneK1K IDs to CPG (internal) and TOB (external) IDs
+    sample_key_df = pd.read_csv(
+        AnyPath(dataset_path(sample_mapping_file)), sep='\t', index_col=0
+    )
+
+    cpg_id = sample_key_df[sample_key_df.index == onek1k_id]['InternalID'].values[0]
+    logging.info(f'CPG ID: {cpg_id}')  # e.g., 'CPG9951'
+
+    # define output filename and check if it already exists
+    output_filename = AnyPath(output_path(f'{cpg_id}_{gene_name}_{window_size}.csv'))
+    logging.info(f'Output file: {output_filename}')
+
+    # skip if file already exists
+    if output_filename.exists():
+        raise Exception(f'File {output_filename} already exists, exiting')
+
+    init_batch()
+    # get VEP-annotated WGS object (hail matrix table)
+    # mt = hl.read_matrix_table('gs://cpg-tob-wgs-test/tob_wgs_vep/v1/vep105_GRCh38.mt')
+    mt = hl.read_matrix_table(MT)
+    logging.info(f'Number of total variants: {mt.count()[0]}')
+
+    # get relevant chromosome
+    gene_df = pd.read_csv(AnyPath(dataset_path(gene_file)), sep='\t', index_col=0)
+    chrom = gene_df[gene_df['gene_name'] == gene_name]['chr']
+
+    # filter to relevant chromosome to speed densification up
+    mt = mt.filter_rows(mt.locus.contig == ('chr'+chrom))  # figure out actual syntax
+    logging.info(f'Number of variants on chromosome {chrom}: {mt.count()[0]}')
+
+    # densify
+    mt = hl.experimental.densify(mt)
+
+    # filter out low QC variants
+    mt = mt.filter_rows(hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)
+    mt_path = output_path('densified_chrom_and_qc_filtered.mt', 'tmp')
+    mt = mt.checkpoint(mt_path, overwrite=True)
+    logging.info(f'Number of QC-passing variants: {mt.count()[0]}')
+
+    # select matrix down to that one donor
+    donor_mt = mt.filter_cols(mt.s == cpg_id)
+
+    # get gene body position (start and end) and build interval
+    # include variants up to {window size} up- and downstream
+    interval_start = int(gene_df[gene_df['gene_name'] == gene_name]['start']) - int(
+        window_size
+    )
+    interval_end = int(gene_df[gene_df['gene_name'] == gene_name]['end']) + int(
+        window_size
+    )
```

Comment:
You'll have to clip these against the chromosome boundaries: https://github.com/populationgenomics/tob-wgs/pull/108 (@MattWellie)

Reply:
reference to this?
https://github.com/populationgenomics/tob-wgs/blob/main/scripts/eqtl_hail_batch/generate_eqtl_spearman.py#L351-L360

Reply:
sorry, could you provide more detail? I thought this worked fine previously?

Reply:
If by extending the interval around the gene you go out of chromosome / contig boundaries, the filtering step below will fail (e.g. imagine you're asking Hail to filter from (negative!) position -100 to 400 or so).

Reply:
ah yes ok that makes sense! Will have a look at Matt's code :)